src/librustc_session/session.rs

```diff
@@ -961,23 +959,18 @@ pub fn build_session_with_source_map(
         DiagnosticOutput::Default => None,
         DiagnosticOutput::Raw(write) => Some(write),
     };
+    let loader = file_loader.unwrap_or(Box::new(RealFileLoader));
+    let source_map = Lrc::new(SourceMap::with_file_loader_and_unset_hash_kind(loader, sopts.file_path_mapping()));
```

Comment:
@eddyb I was able to delay the source map construction until here. It only needs to make it a few more lines.

However there is a circular dependency:
- constructing a SourceMap requires knowing the hash algorithm
- the hash algorithm comes from the Target specification
- reading the Target can fail and produce a Diagnostic
- producing a Diagnostic requires a SourceMap

Is there a good way we can raise a diagnostic without a SourceMap when reading the target information?

Reply:
I resolved this by using an `early_error` for failures parsing the target information.